docs/man/man1/busers.1
.ds ]W %
.ds ]L
.nh
.TH busers 1 "Lava Version 1.0 - Sept 2007"
.br
.SH NAME
\fBbusers\fR - displays information about users and user groups
.SH SYNOPSIS
.BR
.PP
.PP
\fBbusers\fR\fB \fR[\fIuser_name \fR... | \fIuser_group\fR ... | \fBall\fR]\fB 
\fR.PP
\fBbusers\fR\fB \fR[\fB-h\fR | \fB-V\fR]
.SH DESCRIPTION
.BR
.PP
.PP
\fB\fRDisplays information about users and user groups.
.PP
By default, displays information about the user who runs the 
command.
.SH OPTIONS
.BR
.PP
.TP 
\fIuser_name\fR ... |\fB \fR\fIuser_group \fR... |\fB \fR\fBall
\fR
.IP
Displays information about the specified users or user groups, or about 
all users if you specify all.


.TP 
\fB-h
\fR
.IP
Prints command usage to stderr and exits.


.TP 
\fB-V
\fR
.IP
Prints Lava release version to stderr and exits. 


.SH OUTPUT
.BR
.PP
.PP
A listing of the users and user groups is displayed with the following 
fields: 
.SS USER/GROUP 
.BR
.PP
.PP
The name of the user or user group. 
.SS JL/P 
.BR
.PP
.PP
The maximum number of job slots that can be processed 
simultaneously for the specified users on each processor. 
These job slots are used by running and suspended jobs or by pending 
jobs which have jobs slots reserved for them.  This job limit is 
configured per processor so that multiprocessor hosts have more job 
slots. If the dash character (-) is displayed, there is no limit. JL/P is 
defined in the Lava configuration file lsb.users(5). 
.SS MAX 
.BR
.PP
.PP
The maximum number of job slots that can be processed concurrently 
for the specified users' jobs. These job slots are used by running 
and suspended jobs or by pending jobs which have job slots reserved 
for them.  If the character `-' is displayed, there is no limit. MAX 
is defined by the MAX_JOBS parameter in the configuration file 
lsb.users(5). 
.SS NJOBS 
.BR
.PP
.PP
The current number of job slots used by specified users' jobs. A parallel 
job that is pending is counted as \fIn\fR job slots for it will use \fIn\fR job slots 
in the queue when it is dispatched. 
.SS PEND 
.BR
.PP
.PP
The number of pending job slots used by jobs of the specified users.
.SS RUN 
.BR
.PP
.PP
The number of job slots used by running jobs of the specified users. 
.SS SSUSP 
.BR
.PP
.PP
The number of job slots used by the system-suspended jobs of the 
specified users. 
.SS USUSP
.BR
.PP
.PP
The number of job slots used by user-suspended jobs of the specified 
users. 
.SS RSV 
.BR
.PP
.PP
The number of job slots used by pending jobs of the specified users 
which have job slots reserved for them. 
.SH SEE ALSO
.BR
.PP
.PP
bugroup(1), lsb.users(5), lsb.queues(5)


src/lsbatch/cmd/busers.c
/* $Id: busers.c,v 1.2 2007/08/15 22:18:44 tmizan Exp $
 * Copyright (C) 2007 Platform Computing Inc
 * Copyright (C) LavaLite Contributors
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 *
 */

#include "lsbatch/cmd/cmd.h"

#define USR_NAME_LENGTH    18
#define USR_JL_P_LENGTH    5
#define USR_NJOBS_LENGTH   6
#define USR_PEND_LENGTH    6
#define USR_RUN_LENGTH     6
#define USR_SSUSP_LENGTH   6
#define USR_USUSP_LENGTH   6
#define USR_RSV_LENGTH     6
#define USR_MAX_LENGTH     6

static void display_users (struct userInfoEnt *, int);
static void sort_users (struct userInfoEnt *, int);

static char fomt[200];

void
usage (char *cmd)
{
    fprintf(stderr, "Usage");
    fprintf(stderr, ": %s [-h] [-V] [user_name ...] [all]\n", cmd);
    exit(-1);
}

int
main (int argc, char **argv)
{
    extern int optind;
    int cc, numUsers;
    struct userInfoEnt  *usrInfo;
    char **users=NULL, **userPoint;
    int all = FALSE;

    if (lsb_init(argv[0]) < 0) {
	lsb_perror("lsb_init");
	exit(-1);
    }

    while ((cc = getopt(argc, argv, "Vh")) != EOF) {
        switch (cc) {
	case 'V':
	    fprintf(stderr, "%s\n", LAVALITE_VERSION_STR);
	    exit(0);
        case 'h':
        default:
            usage(argv[0]);
        }
    }
    numUsers = getNames (argc, argv, optind, &users, &all, "user");
    if (!all && !numUsers) {
        numUsers = 1;
        userPoint = NULL;
    } else
        userPoint = users;
    if (!(usrInfo = lsb_userinfo (userPoint, &numUsers))) {
        if (lsberrno == LSBE_BAD_USER && userPoint)
            lsb_perror(users[numUsers]);
        else
            lsb_perror(NULL);
        exit (-1);
    }
    display_users(usrInfo, numUsers);
    exit(0);

}

static void
display_users (struct userInfoEnt *reply, int numReply)
{
    int i;
    int first = TRUE;
    char maxJobs[MAX_CHARLEN], procJobLimit[MAX_CHARLEN];

    if (first) {
	first = FALSE;

	if (lsbMode_ & LSB_MODE_BATCH) {
            prtWord(USR_NAME_LENGTH, I18N_USER_GROUP, 0);
            prtWord(USR_JL_P_LENGTH, I18N_JL_P, -1);
            prtWord(USR_MAX_LENGTH,  "MAX", -1);
	}
	else
            prtWord(USR_NAME_LENGTH, "USER", 0);

        prtWord(USR_NJOBS_LENGTH, "NJOBS", -1);
        prtWord(USR_PEND_LENGTH,  "PEND",  -1);
        prtWord(USR_RUN_LENGTH,   "RUN",   -1);
        prtWord(USR_SSUSP_LENGTH, "SSUSP", -1);
        prtWord(USR_USUSP_LENGTH, "USUSP", -1);
        prtWord(USR_RSV_LENGTH,   "RSV",   -1);
        printf("\n");
    }
    if (numReply > 1)
	sort_users (reply, numReply);

    for (i = 0; i < numReply; i++) {

        if (reply[i].procJobLimit < INFINIT_FLOAT) {
            sprintf(fomt, "%%%d.1f ", USR_JL_P_LENGTH);
            sprintf (procJobLimit, fomt, reply[i].procJobLimit);
	}
        else
             strcpy(procJobLimit, prtDash(USR_JL_P_LENGTH) );

        if (reply[i].maxJobs < INFINIT_INT)
            strcpy(maxJobs, prtValue(USR_MAX_LENGTH, reply[i].maxJobs) );
        else
             strcpy(maxJobs, prtDash(USR_MAX_LENGTH) );

        prtWordL(USR_NAME_LENGTH, reply[i].user);

	if (lsbMode_ & LSB_MODE_BATCH) {
            sprintf(fomt, "%%%ds%%%ds", USR_JL_P_LENGTH,
                                        USR_MAX_LENGTH );
	    printf(fomt,
                   procJobLimit, maxJobs);
	};

	if (strcmp(reply[i].user, "default") == 0) {
            printf("%s%s%s%s%s%s\n",
                   prtDash(USR_NJOBS_LENGTH), prtDash(USR_PEND_LENGTH),
                   prtDash(USR_RUN_LENGTH),   prtDash(USR_SSUSP_LENGTH),
                   prtDash(USR_USUSP_LENGTH), prtDash(USR_RSV_LENGTH)
                   );
	}
	else if (reply[i].numJobs == -INFINIT_INT) {
            printf("%s%s%s%s%s%s\n",
                   prtDash(USR_NJOBS_LENGTH), prtDash(USR_PEND_LENGTH),
                   prtDash(USR_RUN_LENGTH),   prtDash(USR_SSUSP_LENGTH),
                   prtDash(USR_USUSP_LENGTH), prtDash(USR_RSV_LENGTH)
                   );
	}
        else {
        sprintf(fomt, "%%%dd %%%dd %%%dd %%%dd %%%dd %%%dd\n",
                                                  USR_NJOBS_LENGTH,
                                                  USR_PEND_LENGTH,
                                                  USR_RUN_LENGTH,
                                                  USR_SSUSP_LENGTH,
                                                  USR_USUSP_LENGTH,
                                                  USR_RSV_LENGTH );
	printf(fomt,
                reply[i].numJobs, reply[i].numPEND, reply[i].numRUN,
                reply[i].numSSUSP, reply[i].numUSUSP, reply[i].numRESERVE);
	};
    }

}

static void
sort_users (struct userInfoEnt *users, int numUsers)
{
    int i, j, k;
    struct userInfoEnt temUser;

    for (k = numUsers / 2; k > 0; k /= 2) {
        for (i = k; i < numUsers; i++) {
            for (j = i-k; j >= 0; j -= k) {
                 if (strcmp(users[j].user, users[j+k].user) < 0)
                     break;
                 memcpy(&temUser, users+j, sizeof (struct userInfoEnt));
                 memcpy(users+j, users+j+k, sizeof (struct userInfoEnt));
                 memcpy(users+j+k, &temUser, sizeof (struct userInfoEnt));
            }
        }
    }
    return;

}
